Cnsfileaccessconfig controller implementation

chethanv28 opened the pull request:
<!-- Thanks for sending a pull request! -->

**What this PR does / why we need it**:
This PR is implementing Cnsfileaccessconfig controller
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
Cnsfileaccessconfig controller implementation
```

RaunakShah commented on pkg/syncer/cnsoperator/controller/cnsfileaccessconfig/cnsfileaccessconfig_controller.go:
```diff
@@ -0,0 +1,343 @@
+			}
+		}
+		// Update finalizer and ownerRef on CnsFileAccessConfig instance
+		if !cnsFinalizerExists {
```
This assumes that if CNSFinalizer exists on the instance then OwnerRef does too. Are we sure about this assumption?

chethanv28 replied:
Looks like the comment is wrong. Updated

RaunakShah commented on pkg/syncer/cnsoperator/controller/cnsfileaccessconfig/util.go:
```diff
@@ -0,0 +1,108 @@
+}
+
+// setInstanceOwnerRef sets ownerRef on CnsFileAccessConfig instance to VM instance
+func setInstanceOwnerRef(instance *cnsfileaccessconfigv1alpha1.CnsFileAccessConfig, vmName string,
```
Can this function be moved to `cnsoperator/util/util.go`? I see that other controller use/can use this too.

chethanv28 replied:
Created this issue: https://github.com/kubernetes-sigs/vsphere-csi-driver/issues/587 for tracking the suggestion.
I am not addressing the review comment in this PR to avoid more complexity/testing cycles. It would need some additional effort to refactor 4 different Custom resource controller.go files and also run some tests.
Thanks

RaunakShah commented on pkg/syncer/cnsoperator/controller/cnsfileaccessconfig/cnsfileaccessconfig_controller.go:
```diff
@@ -0,0 +1,343 @@
+	return reconcile.Result{}, nil
+}
+
+// setInstanceSuccess sets instance to success and records an event on the CnsRegisterVolume instance
```
Cant all the functions from here on be moved to `cnsoperator/util/util.go`? I see that other controllers use the same logic, the only difference is the `instance` type. Have you tried using `runtime.Object` as a parameter instead of the exact type, and then cast it accordingly? All controllers can then leverage the same functions.

chethanv28 replied:
Created this issue: https://github.com/kubernetes-sigs/vsphere-csi-driver/issues/587 for tracking the suggestion.
I am not addressing the review comment in this PR to avoid more complexity/testing cycles. It would need some additional effort to refactor 4 different Custom resource controller.go files and also run some tests.
Thanks

gohilankit replied:
CnsRegisterVolume -> CnsFileAccessConfig in method doc.

RaunakShah commented on pkg/syncer/cnsoperator/controller/cnsfileaccessconfig/cnsfileaccessconfig_controller.go:
```diff
@@ -0,0 +1,343 @@
+			return reconcile.Result{RequeueAfter: timeout}, nil
+		}
+
+		// Query volume
```
This is not part of the design, but i'm wondering whether its worth/possible to first query the API server to get the access point (if it exists) and fall back to CNS if it doesnt. There was a change to CNS QueryVolume recently that drops concurrent requests beyond a certain limit

chethanv28 replied:
can you create a separate issue to track this please. Thanks

RaunakShah commented on pkg/syncer/cnsoperator/controller/cnsfileaccessconfig/cnsfileaccessconfig_controller.go:
```diff
@@ -0,0 +1,343 @@
+		}
+
+		// Query volume
+		log.Infof("Querying volume: %s for CnsFileAccessConfig request with name: %q on namespace: %q",
```
Debug log?

chethanv28 replied:
+1

RaunakShah commented on pkg/syncer/cnsoperator/controller/cnsfileaccessconfig/cnsfileaccessconfig_controller.go:
```diff
@@ -0,0 +1,343 @@
+				return reconcile.Result{RequeueAfter: timeout}, nil
+			}
+			log.Errorf("Failed to query CNS volume: %s with error: %+v", volumeID, err)
+			setInstanceError(ctx, r, instance, "Unable to find the volume in CNS")
```
Can you add the error returned by CNS to the instance as well?

chethanv28 replied:
Done

gohilankit replied:
Shouldn't this be a different error message? This  looks pretty much like the one when CNS volume is not found. May be you can set the instance error as "Failed to query CNS volume".

RaunakShah commented on pkg/syncer/cnsoperator/controller/cnsfileaccessconfig/cnsfileaccessconfig_controller.go:
```diff
@@ -0,0 +1,343 @@
+			if err != nil {
+				msg := fmt.Sprintf("failed to update CnsFileAccessConfig instance: %q on namespace: %q. Error: %+v",
+					instance.Name, instance.Namespace, err)
+				recordEvent(ctx, r, instance, v1.EventTypeWarning, msg)
```
Why do we use `recordEvent` in some places and `setInstanceSuccess/Err` in others?

chethanv28 replied:
`recordEvent` : records the event and also sets the backOffDuration for the instance
It is invoked explicitly at 2 places: One(1) is for recording a Get failure and one(2) while updating finalizers.

Otherwise `recordEvent` is called inside setInstanceSuccess/setInstanceErr.
And the two explicit invocations have nothing to do with instance.Spec or instance.Status. 
I do not see anything incorrect here. PTAL

RaunakShah commented on pkg/syncer/cnsoperator/controller/cnsfileaccessconfig/cnsfileaccessconfig_controller.go:
```diff
@@ -0,0 +1,343 @@
+		}
+		if len(accessPoints) == 0 {
+			log.Errorf("No access points found for volume: %q", volumeID)
+			setInstanceError(ctx, r, instance, "No access points found.")
```
Can you elaborate on the message?

chethanv28 replied:
Done

RaunakShah commented on pkg/syncer/cnsoperator/controller/cnsfileaccessconfig/cnsfileaccessconfig_controller.go:
```diff
@@ -0,0 +1,343 @@
+		}
+
+		// Read the CnsFileAccessConfig instance again because the instance is already modified
+		err = r.client.Get(ctx, request.NamespacedName, instance)
```
What about in other places between here and update on line 198? There are cases where the error is set on the object, so isnt there a need to `get()` before that too?

chethanv28 replied:
Not needed, since other places are only changing the status(setInstanceError)

RaunakShah replied:
`client.Update` modifies the input instance according to the documentation:

```
	// Update updates the given obj in the Kubernetes cluster. obj must be a
	// struct pointer so that obj can be updated with the content returned by the Server.
```

In that case, this `Get()` is redundant right?

RaunakShah commented on pkg/syncer/cnsoperator/controller/cnsfileaccessconfig/cnsfileaccessconfig_controller.go:
```diff
@@ -0,0 +1,343 @@
+			recordEvent(ctx, r, instance, v1.EventTypeWarning, msg)
+			return reconcile.Result{RequeueAfter: timeout}, nil
+		}
+
```
Can you add a TODO here for getting the network configuration details and configuring File share ACLs?

chethanv28 replied:
Done

RaunakShah commented on pkg/syncer/cnsoperator/controller/cnsfileaccessconfig/cnsfileaccessconfig_controller.go:
```diff
@@ -0,0 +1,343 @@
+	}
+
+	if instance.DeletionTimestamp != nil {
+		removeFinalizerFromCRDInstance(ctx, instance)
```
Can you add a TODO here for actions to be taken before removing finalizer?

chethanv28 replied:
Added

RaunakShah commented on pkg/syncer/cnsoperator/controller/cnsfileaccessconfig/util.go:
```diff
@@ -0,0 +1,108 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	apitypes "k8s.io/apimachinery/pkg/types"
+	cnsfileaccessconfigv1alpha1 "sigs.k8s.io/vsphere-csi-driver/pkg/apis/cnsoperator/cnsfileaccessconfig/v1alpha1"
```
Add a new line?

chethanv28 replied:
Ok

RaunakShah commented on pkg/syncer/cnsoperator/controller/cnsfileaccessconfig/util.go:
```diff
@@ -0,0 +1,108 @@
+// getVirtualMachine gets the virtual machine instance with a name on a SV namespace
+func getVirtualMachine(ctx context.Context, vmName string, namespace string) (*vmoperatortypes.VirtualMachine, error) {
+	log := logger.GetLogger(ctx)
+	restClientConfig, err := k8s.GetKubeConfig(ctx)
```
Instead of initializing this for every call, can we add `vmOperatorClient` as a member to `ReconcileCnsFileAccessConfig struct`

chethanv28 replied:
Done

chethanv28 commented on pkg/syncer/cnsoperator/controller/cnsfileaccessconfig/cnsfileaccessconfig_controller.go:
```diff
@@ -0,0 +1,343 @@
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	apis "sigs.k8s.io/vsphere-csi-driver/pkg/apis/cnsoperator"
```
Done

chethanv28 commented on pkg/syncer/cnsoperator/controller/cnsfileaccessconfig/cnsfileaccessconfig_controller.go:
```diff
@@ -0,0 +1,343 @@
+	defaultMaxWorkerThreadsForFileAccessConfig = 10
+)
+
+// backOffDuration is a map of cnsnodevmattachment name's to the time after which a request
```
Done

chethanv28 commented on pkg/syncer/cnsoperator/controller/cnsfileaccessconfig/cnsfileaccessconfig_controller.go:
```diff
@@ -0,0 +1,343 @@
+		return err
+	}
+
+	// eventBroadcaster broadcasts events on cnsnodevmattachment instances to the event sink
```
Done

chethanv28 reviewed (commented):
Addressed review comments from @RaunakShah

gohilankit commented on pkg/syncer/cnsoperator/util/util.go:
```diff
@@ -0,0 +1,49 @@
+	log := logger.GetLogger(ctx)
+	// Get PVC by pvcName from namespace
+	pvc := &v1.PersistentVolumeClaim{}
+	err := client.Get(ctx, k8stypes.NamespacedName{Name: pvcName, Namespace: namespace}, pvc)
```
Does this fetch object from API server? If yes, then we are hitting API server twice in this method.
Can this be optimized to use shared informer's PV/PVC listers? If it's not found using PVCLister, then we query the API server.

chethanv28 replied:
I agree with this improvement. However I am creating a github issue for adding informers.

chethanv28 replied:
Issue number 599

gohilankit replied:
Can you add a TODO here with a link to the issue?

chethanv28 replied:
Done, added

gohilankit commented on pkg/syncer/cnsoperator/controller/cnsfileaccessconfig/cnsfileaccessconfig_controller.go:
```diff
@@ -0,0 +1,372 @@
+		if finalizer == cnsoperatortypes.CNSFinalizer {
+			log.Debugf("Removing %q finalizer from CnsFileAccessConfig instance with name: %q on namespace: %q",
+				cnsoperatortypes.CNSFinalizer, instance.Name, instance.Namespace)
+			instance.Finalizers = append(instance.Finalizers[:i], instance.Finalizers[i+1:]...)
```
nit: `break` from the loop once you have remove the finalizer.

gohilankit commented on pkg/syncer/cnsoperator/controller/cnsfileaccessconfig/util.go:
```diff
@@ -0,0 +1,95 @@
+			log.Warnf("Maximum number of worker threads to run set in env variable WORKER_THREADS_FILE_ACCESS_CONFIG %s is invalid, will use the default value %d", v, defaultMaxWorkerThreadsForFileAccessConfig)
+		}
+	} else {
+		log.Debugf("WORKER_THREADS_REGISTER_VOLUME is not set. Picking the default value %d", defaultMaxWorkerThreadsForFileAccessConfig)
```
WORKER_THREADS_REGISTER_VOLUME  -> WORKER_THREADS_FILE_ACCESS_CONFIG

chethanv28 replied:
changed

gohilankit commented on pkg/syncer/cnsoperator/controller/cnsfileaccessconfig/util.go:
```diff
@@ -0,0 +1,95 @@
+					v, defaultMaxWorkerThreadsForFileAccessConfig, defaultMaxWorkerThreadsForFileAccessConfig)
+			} else {
+				workerThreads = value
+				log.Debugf("Maximum number of worker threads to run to reconcile CnsRegisterVolume instances is set to %d", workerThreads)
```
CnsRegisterVolume -> CnsFileAccessConfig

chethanv28 replied:
Done

chethanv28 commented:
Created 2 PVCs
```
root@4230b5ef311c6c8686e63f62f5d303c4 [ ~ ]# kubectl get pvc -A
NAMESPACE   NAME                         STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS         AGE
default     example-vanilla-file-pvc     Bound    pvc-4fc254f6-91aa-478e-9872-afa44a8c0f7b   1Mi        RWX            gc-storage-profile   56m
default     example-vanilla-file-pvc-1   Bound    pvc-15799762-45b6-4ff1-9d63-36f39f8aaa87   1Mi        RWX            gc-storage-profile   34m
```
Created 6 pods
```
root@4230b5ef311c6c8686e63f62f5d303c4 [ ~ ]# kubectl get pods -o wide
NAME                         READY   STATUS    RESTARTS   AGE     IP             NODE                                                     NOMINATED NODE   READINESS GATES
example-vanilla-file-pod-1   1/1     Running   0          20s     192.0.254.12   test-cluster-e2e-script-workers-hggkp-6bb876f4df-tx826   <none>           <none>
example-vanilla-file-pod-2   1/1     Running   0          20s     192.0.254.13   test-cluster-e2e-script-workers-hggkp-6bb876f4df-tx826   <none>           <none>
example-vanilla-file-pod-3   1/1     Running   0          20s     192.0.254.14   test-cluster-e2e-script-workers-hggkp-6bb876f4df-tx826   <none>           <none>
example-vanilla-file-pod-4   1/1     Running   0          29m     192.0.254.9    test-cluster-e2e-script-workers-hggkp-6bb876f4df-tx826   <none>           <none>
example-vanilla-file-pod-5   1/1     Running   0          29m     192.0.254.10   test-cluster-e2e-script-workers-hggkp-6bb876f4df-tx826   <none>           <none>
example-vanilla-file-pod1    1/1     Running   0          5m30s   192.0.254.11   test-cluster-e2e-script-workers-hggkp-6bb876f4df-tx826   <none>           <none>
```
Pod - PVC mapping
```
example-vanilla-file-pod1     -------------> example-vanilla-file-pvc
example-vanilla-file-pod-1   -------------> example-vanilla-file-pvc
example-vanilla-file-pod-2  -------------> example-vanilla-file-pvc
example-vanilla-file-pod-3  -------------> example-vanilla-file-pvc

example-vanilla-file-pod-4  -------------> example-vanilla-file-pvc-1
example-vanilla-file-pod-5  -------------> example-vanilla-file-pvc-1
```

Verified that CnsFileAccessConfigController is created and accesspoints are available:

```
root@4230aca1613dcd75798cb2b100c75710 [ ~ ]# kubectl describe cnsfileaccessconfigs.cns.vmware.com -n test-gc-e2e-demo-ns 
Name:         test-cluster-e2e-script-workers-hggkp-6bb876f4df-tx826-b765ab11-d287-403f-b31b-16d057e5ee9f-15799762-45b6-4ff1-9d63-36f39f8aaa87
Namespace:    test-gc-e2e-demo-ns
Labels:       <none>
Annotations:  <none>
API Version:  cns.vmware.com/v1alpha1
Kind:         CnsFileAccessConfig
Metadata:
  Creation Timestamp:  2021-01-18T20:21:27Z
  Generation:          2
  Managed Fields:
    API Version:  cns.vmware.com/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:spec:
        .:
        f:pvcName:
        f:vmName:
      f:status:
    Manager:      vsphere-csi
    Operation:    Update
    Time:         2021-01-18T20:21:27Z
    API Version:  cns.vmware.com/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:status:
        f:accessPoints:
          .:
          f:NFSv4.1:
        f:done:
    Manager:         kubectl-edit
    Operation:       Update
    Time:            2021-01-18T20:23:25Z
  Resource Version:  2813134
  Self Link:         /apis/cns.vmware.com/v1alpha1/namespaces/test-gc-e2e-demo-ns/cnsfileaccessconfigs/test-cluster-e2e-script-workers-hggkp-6bb876f4df-tx826-b765ab11-d287-403f-b31b-16d057e5ee9f-15799762-45b6-4ff1-9d63-36f39f8aaa87
  UID:               6b1375c6-9faf-46c5-a71c-f94e989005e5
Spec:
  Pvc Name:  b765ab11-d287-403f-b31b-16d057e5ee9f-15799762-45b6-4ff1-9d63-36f39f8aaa87
  Vm Name:   test-cluster-e2e-script-workers-hggkp-6bb876f4df-tx826
Status:
  Access Points:
    NFSv4.1:  172.16.10.10:/vsanfs/5253aad5-6f65-4679-7ee1-d8e7a40bb3d9
  Done:       true
Events:       <none>


Name:         test-cluster-e2e-script-workers-hggkp-6bb876f4df-tx826-b765ab11-d287-403f-b31b-16d057e5ee9f-4fc254f6-91aa-478e-9872-afa44a8c0f7b
Namespace:    test-gc-e2e-demo-ns
Labels:       <none>
Annotations:  <none>
API Version:  cns.vmware.com/v1alpha1
Kind:         CnsFileAccessConfig
Metadata:
  Creation Timestamp:  2021-01-18T20:45:07Z
  Generation:          2
  Managed Fields:
    API Version:  cns.vmware.com/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:spec:
        .:
        f:pvcName:
        f:vmName:
      f:status:
    Manager:      vsphere-csi
    Operation:    Update
    Time:         2021-01-18T20:45:07Z
    API Version:  cns.vmware.com/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:status:
        f:accessPoints:
          .:
          f:NFSv4.1:
        f:done:
    Manager:         kubectl-edit
    Operation:       Update
    Time:            2021-01-18T20:46:21Z
  Resource Version:  2822832
  Self Link:         /apis/cns.vmware.com/v1alpha1/namespaces/test-gc-e2e-demo-ns/cnsfileaccessconfigs/test-cluster-e2e-script-workers-hggkp-6bb876f4df-tx826-b765ab11-d287-403f-b31b-16d057e5ee9f-4fc254f6-91aa-478e-9872-afa44a8c0f7b
  UID:               bee1bdb2-9ce6-402a-b12f-6437e7fc3751
Spec:
  Pvc Name:  b765ab11-d287-403f-b31b-16d057e5ee9f-4fc254f6-91aa-478e-9872-afa44a8c0f7b
  Vm Name:   test-cluster-e2e-script-workers-hggkp-6bb876f4df-tx826
Status:
  Access Points:
    NFSv4.1:  172.16.10.10:/vsanfs/520ff9f1-7c41-3209-b044-9408910db08d
  Done:       true
Events:       <none>
```

chethanv28 commented on pkg/syncer/cnsoperator/controller/cnsfileaccessconfig/cnsfileaccessconfig_controller.go:
```diff
@@ -0,0 +1,372 @@
+}
+
+// blank assignment to verify that ReconcileCnsFileAccessConfig implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileCnsFileAccessConfig{}
```
Looks like it is needed and this is used in all the CRD controllers. keeping it same as other existing controllers

chethanv28 commented on pkg/syncer/cnsoperator/controller/cnsfileaccessconfig/cnsfileaccessconfig_controller.go:
```diff
@@ -0,0 +1,372 @@
+	ctx, log := logger.GetNewContextWithLogger()
+	// Fetch the CnsFileAccessConfig instance
+	instance := &cnsfileaccessconfigv1alpha1.CnsFileAccessConfig{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
```
Removed

gohilankit commented:
I don't see the `owner references` being set in your test results. Can you confirm?

chethanv28 commented:
> I don't see `ownerReferences` being set in your test results. Can you confirm?

In the previous instance I had forgotten to add rbac rules while testing. Here is the updated testing result with all rbac rules. Thanks for pointing out though.

```
root@4230b5ef311c6c8686e63f62f5d303c4 [ ~ ]# kubectl describe cnsfileaccessconfigs.cns.vmware.com test-cluster-e2e-script-workers-hggkp-6bb876f4df-tx826-b765ab11-d287-403f-b31b-16d057e5ee9f-606679a3-02e7-4bc6-8835-c14a1364f0d6 -n test-gc-e2e-demo-ns
Name:         test-cluster-e2e-script-workers-hggkp-6bb876f4df-tx826-b765ab11-d287-403f-b31b-16d057e5ee9f-606679a3-02e7-4bc6-8835-c14a1364f0d6
Namespace:    test-gc-e2e-demo-ns
Labels:       <none>
Annotations:  <none>
API Version:  cns.vmware.com/v1alpha1
Kind:         CnsFileAccessConfig
Metadata:
  Creation Timestamp:  2021-01-19T04:39:14Z
  Finalizers:
    cns.vmware.com
  Generation:  2
  Managed Fields:
    API Version:  cns.vmware.com/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:spec:
        .:
        f:pvcName:
        f:vmName:
      f:status:
    Manager:      vsphere-csi
    Operation:    Update
    Time:         2021-01-19T04:39:14Z
    API Version:  cns.vmware.com/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:finalizers:
          .:
          v:"cns.vmware.com":
        f:ownerReferences:
          .:
          k:{"uid":"fd78e879-07d1-4646-b19b-172d88a6dd5e"}:
            .:
            f:apiVersion:
            f:blockOwnerDeletion:
            f:controller:
            f:kind:
            f:name:
            f:uid:
      f:status:
        f:accessPoints:
          .:
          f:NFSv3:
          f:NFSv4.1:
        f:done:
    Manager:    vsphere-syncer
    Operation:  Update
    Time:       2021-01-19T04:39:16Z
  Owner References:
    API Version:           v1
    Block Owner Deletion:  true
    Controller:            true
    Kind:                  VirtualMachine
    Name:                  test-cluster-e2e-script-workers-hggkp-6bb876f4df-tx826
    UID:                   fd78e879-07d1-4646-b19b-172d88a6dd5e
  Resource Version:        3020257
  Self Link:               /apis/cns.vmware.com/v1alpha1/namespaces/test-gc-e2e-demo-ns/cnsfileaccessconfigs/test-cluster-e2e-script-workers-hggkp-6bb876f4df-tx826-b765ab11-d287-403f-b31b-16d057e5ee9f-606679a3-02e7-4bc6-8835-c14a1364f0d6
  UID:                     31dee6ef-a11b-4cfd-a966-c6e2268221d8
Spec:
  Pvc Name:  b765ab11-d287-403f-b31b-16d057e5ee9f-606679a3-02e7-4bc6-8835-c14a1364f0d6
  Vm Name:   test-cluster-e2e-script-workers-hggkp-6bb876f4df-tx826
Status:
  Access Points:
    NFSv3:    h172-16-10-10.chethanvtest.testdomain:/52096f4c-5ce1-f8cc-cf7f-6637edf2c3ae
    NFSv4.1:  h172-16-10-10.chethanvtest.testdomain:/vsanfs/52096f4c-5ce1-f8cc-cf7f-6637edf2c3ae
  Done:       true
Events:
  Type    Reason                        Age   From            Message
  ----    ------                        ----  ----            -------
  Normal  CnsFileAccessConfigSucceeded  19s   cns.vmware.com  Successfully configured access points of VM: "test-cluster-e2e-script-workers-hggkp-6bb876f4df-tx826" on the volume: "b765ab11-d287-403f-b31b-16d057e5ee9f-606679a3-02e7-4bc6-8835-c14a1364f0d6"
```

RaunakShah commented on pkg/syncer/cnsoperator/controller/cnsfileaccessconfig/cnsfileaccessconfig_controller.go:
```diff
@@ -0,0 +1,374 @@
+		if !cnsFinalizerExists {
+			// Add finalizer.
+			instance.Finalizers = append(instance.Finalizers, cnsoperatortypes.CNSFinalizer)
+			err = updateCnsFileAccessConfig(ctx, r.client, instance)
```
Shouldn't this be outside of this if statement? I think setting the ownerRef and Finalizer are pre requisites before we attempt to reconcile CnsFileAccessConfig instances

chethanv28 replied:
Addressed this comment in the latest commit

chethanv28 replied:
As discussed offline, I have updated the PR with a 2nd commit to address this concern.

gohilankit commented:
Changes look good to me.
/approve

RaunakShah commented on pkg/syncer/cnsoperator/controller/cnsfileaccessconfig/cnsfileaccessconfig_controller.go:
```diff
@@ -0,0 +1,374 @@
+		log.Info(msg)
+	}
+
+	if instance.DeletionTimestamp != nil {
```
This is subjective - Can you move this check to line 177? Then the checks on line 179 and 188 do not need to verify that `instance.DeletionTimestamp == nil` in addition to `instance.Status.Done`. Makes the code easier to read IMO.

chethanv28 replied:
Addressed this comment in the latest commit

chethanv28 commented:
Testing after addressing comments:

```
root@423027f30a593bc7f5961c33400618ce [ ~ ]# kubectl describe cnsfileaccessconfigs.cns.vmware.com -n test-gc-e2e-demo-ns test-cluster-e2e-script-workers-hggkp-6bb876f4df-tx826-b765ab11-d287-403f-b31b-16d057e5ee9f-c2531eca-02e2-4dc7-b502-7f2a53230a90
Name:         test-cluster-e2e-script-workers-hggkp-6bb876f4df-tx826-b765ab11-d287-403f-b31b-16d057e5ee9f-c2531eca-02e2-4dc7-b502-7f2a53230a90
Namespace:    test-gc-e2e-demo-ns
Labels:       <none>
Annotations:  <none>
API Version:  cns.vmware.com/v1alpha1
Kind:         CnsFileAccessConfig
Metadata:
  Creation Timestamp:  2021-01-19T18:05:14Z
  Finalizers:
    cns.vmware.com
  Generation:  2
  Managed Fields:
    API Version:  cns.vmware.com/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:spec:
        .:
        f:pvcName:
        f:vmName:
      f:status:
    Manager:      vsphere-csi
    Operation:    Update
    Time:         2021-01-19T18:05:14Z
    API Version:  cns.vmware.com/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:finalizers:
          .:
          v:"cns.vmware.com":
        f:ownerReferences:
          .:
          k:{"uid":"fd78e879-07d1-4646-b19b-172d88a6dd5e"}:
            .:
            f:apiVersion:
            f:blockOwnerDeletion:
            f:controller:
            f:kind:
            f:name:
            f:uid:
      f:status:
        f:accessPoints:
          .:
          f:NFSv3:
          f:NFSv4.1:
        f:done:
    Manager:    vsphere-syncer
    Operation:  Update
    Time:       2021-01-19T18:05:15Z
  Owner References:
    API Version:           v1
    Block Owner Deletion:  true
    Controller:            true
    Kind:                  VirtualMachine
    Name:                  test-cluster-e2e-script-workers-hggkp-6bb876f4df-tx826
    UID:                   fd78e879-07d1-4646-b19b-172d88a6dd5e
  Resource Version:        3357749
  Self Link:               /apis/cns.vmware.com/v1alpha1/namespaces/test-gc-e2e-demo-ns/cnsfileaccessconfigs/test-cluster-e2e-script-workers-hggkp-6bb876f4df-tx826-b765ab11-d287-403f-b31b-16d057e5ee9f-c2531eca-02e2-4dc7-b502-7f2a53230a90
  UID:                     fe6981f8-cec9-4baf-81a9-76d631e4816f
Spec:
  Pvc Name:  b765ab11-d287-403f-b31b-16d057e5ee9f-c2531eca-02e2-4dc7-b502-7f2a53230a90
  Vm Name:   test-cluster-e2e-script-workers-hggkp-6bb876f4df-tx826
Status:
  Access Points:
    NFSv3:    h172-16-10-11.chethanvtest.testdomain:/52abe0df-c5ba-2a79-9133-de5dbff538ca
    NFSv4.1:  h172-16-10-10.chethanvtest.testdomain:/vsanfs/52abe0df-c5ba-2a79-9133-de5dbff538ca
  Done:       true
Events:
  Type    Reason                        Age   From            Message
  ----    ------                        ----  ----            -------
  Normal  CnsFileAccessConfigSucceeded  17s   cns.vmware.com  Successfully configured access points of VM: "test-cluster-e2e-script-workers-hggkp-6bb876f4df-tx826" on the volume: "b765ab11-d287-403f-b31b-16d057e5ee9f-c2531eca-02e2-4dc7-b502-7f2a53230a90"
```

@RaunakShah I have addressed the comments. PTAL

RaunakShah commented on pkg/syncer/cnsoperator/controller/cnsfileaccessconfig/cnsfileaccessconfig_controller.go:
```diff
@@ -0,0 +1,353 @@
+
+	// If the CnsFileAccessConfig instance is already successful,
+	// and not deleted by the user, remove the instance from the queue.
+	if instance.Status.Done && instance.DeletionTimestamp == nil {
```
`instance.DeletionTimestamp == nil` can be removed right?

chethanv28 replied:
Makes sense, since we moved the deletion timestamp check to top of the file

RaunakShah commented on pkg/syncer/cnsoperator/controller/cnsfileaccessconfig/cnsfileaccessconfig_controller.go:
```diff
@@ -0,0 +1,353 @@
+	}
+
+	log.Infof("Reconciling CnsFileAccessConfig with instance: %q from namespace: %q. timeout %q seconds", instance.Name, instance.Namespace, timeout)
+	if !instance.Status.Done && instance.DeletionTimestamp == nil {
```
`instance.DeletionTimestamp == nil` can be removed?

chethanv28 replied:
Makes sense, since we moved the deletion timestamp check to top of the file

RaunakShah commented on pkg/syncer/cnsoperator/controller/cnsfileaccessconfig/cnsfileaccessconfig_controller.go:
```diff
@@ -0,0 +1,353 @@
+		setInstanceOwnerRef(instance, instance.Spec.VMName, vm.UID)
+
+		// Add finalizer.
+		instance.Finalizers = append(instance.Finalizers, cnsoperatortypes.CNSFinalizer)
```
CNS Operator finalizer only needs to be added if it doesnt exist right?

chethanv28 replied:
This was there originally, I think I removed while addressing some lint checks. Added back the finalizer check.

RaunakShah commented:
Looks good, thanks for addressing all comments!

/approve
/lgtm
